pkg/scheduler/plugins/elasticquota/core/group_quota_manager.go

```diff
@@ -0,0 +1,658 @@
+		quotaTreeWrapperMap:              make(map[string]*QuotaTreeWrapper),
+		quotaTopoNode:                    make(map[string]*QuotaTopoNode),
+		scaleMinQuotaWhenOverRootResInfo: NewScaleMinQuotaWhenOverRootResInfo(),
+	}
+	quotaManager.quotaInfoMap[extension.SystemQuotaName] = NewQuotaInfo(true, true, extension.SystemQuotaName, extension.SystemQuotaName)
+	quotaManager.quotaInfoMap[extension.SystemQuotaName].setMaxQuotaNoLock(systemGroupMax)
+	quotaManager.quotaInfoMap[extension.DefaultQuotaName] = NewQuotaInfo(true, true, extension.DefaultQuotaName, extension.DefaultQuotaName)
+	quotaManager.quotaInfoMap[extension.DefaultQuotaName].setMaxQuotaNoLock(defaultGroupMax)
+	quotaManager.quotaTreeWrapperMap[extension.RootQuotaName] = NewQuotaTreeWrapper(extension.RootQuotaName)
+	return quotaManager
+}
+
+func (gqm *GroupQuotaManager) SetScaleMinQuotaEnabled(flag bool) {
+	gqm.Lock.Lock()
+	defer gqm.Lock.Unlock()
+
+	gqm.scaleMinQuotaEnabled = flag
+	klog.V(3).Infof("Set ScaleMinQuotaEnabled, flag:%v", gqm.scaleMinQuotaEnabled)
+}
+
+func (gqm *GroupQuotaManager) StartOnce() {
+	gqm.once.Do(func() {
+		go wait.Until(gqm.DumpQuotaInfo, DumpQuotaInfoCycle, wait.NeverStop)
+	})
+}
+
+/***************************************************multi quota runtime process****************************************/
+
+func (gqm *GroupQuotaManager) UpdateClusterTotalResource(deltaRes v1.ResourceList) {
+	gqm.Lock.Lock()
+	defer gqm.Lock.Unlock()
+
+	klog.V(3).Infof("UpdateClusterResource deltaRes:%v", deltaRes)
+	gqm.updateClusterTotalResourceNoLock(deltaRes)
+}
+
+func (gqm *GroupQuotaManager) updateClusterTotalResourceNoLock(deltaRes v1.ResourceList) {
+	gqm.totalResource = quotav1.Add(gqm.totalResource, deltaRes)
+
+	sysAndDefaultUsed := quotav1.Add(gqm.quotaInfoMap[extension.DefaultQuotaName].GetUsed(), gqm.quotaInfoMap[extension.SystemQuotaName].GetUsed())
+	totalResNoSysOrDefault := quotav1.Subtract(gqm.totalResource, sysAndDefaultUsed)
+
+	diffRes := quotav1.Subtract(totalResNoSysOrDefault, gqm.totalResourceIndeed)
+
+	if !quotav1.IsZero(diffRes) {
+		gqm.totalResourceIndeed = totalResNoSysOrDefault.DeepCopy()
+		gqm.quotaTreeWrapperMap[extension.RootQuotaName].SetClusterTotalResource(totalResNoSysOrDefault)
+		klog.V(3).Infof("UpdateClusterResource finish totalResourceIndeed:%v", gqm.totalResourceIndeed)
+	}
+}
+
+func (gqm *GroupQuotaManager) GetClusterTotalResource() v1.ResourceList {
+	gqm.Lock.Lock()
+	defer gqm.Lock.Unlock()
+
+	return gqm.totalResource.DeepCopy()
+}
+
+func (gqm *GroupQuotaManager) UpdateGroupDeltaRequest(quotaName string, deltaReq v1.ResourceList) {
+	gqm.Lock.RLock()
+	defer gqm.Lock.RUnlock()
+
+	gqm.updateGroupDeltaRequestNoLock(quotaName, deltaReq)
+}
+
+// updateGroupDeltaRequestNoLock no need lock gqm.lock
+func (gqm *GroupQuotaManager) updateGroupDeltaRequestNoLock(quotaName string, deltaReq v1.ResourceList) {
+	curToAllParInfos := gqm.getCurToAllParentGroupQuotaInfoNoLock(quotaName)
```

Comment:
`curToAllPar` means `the current node and the current node's parent and ancestor nodes`?

Reply:
yes